### ISSUE TITLE
[BUG] SageMaker deploy fails due to invalid link to model

### ISSUE THREAD

**Issue by Tradunsky** [mlflow/mlflow#2573]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: Nope
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac
- **MLflow installed from (source or binary)**: yep
- **MLflow version (run ``mlflow --version``)**: 0.9.1
- **Python version**: 2.7
- **Exact command to reproduce**: mlflow sagemaker deploy --app-name super-model --model-path model --execution-role-arn arn:aws:iam::<account_id>:role/service-role/AmazonSageMaker-ExecutionRole --run-id d02fd1b8c16d4efa93a85a0510cd1b12 --image-url ecr.amazonaws.com/mlflow:0.9.1 --mode replace --instance-type ml.t2.xlarge --instance-count 1 --vpc-config vpc_config_us_west_1.json --async --archive **--bucket bucket_in_us_east_1 --region-name us-west-1** 

### Describe the problem
When SageMaker endpoint starts it fails with an error:
`Failed to download model data for container \\\“container_1\\\” from URL: \\\“https://s3.us-west-1.amazonaws.com/bucket_in_us_east_1/d02fd1b8c16d4efa93a85a0510cd1b12/model/model.tar.gz\\\“. Please ensure that there is an object located at the URL and that the role passed to CreateModel has permissions to download the object.\“,\“region\“:\“us-west-1\“}”`

So in fact the bucket **bucket_in_us_east_1** does exist, but in different region us-east-1 whereas specified us-west-1.

### Other info / logs
I think the root cause of the error is that mlflow doesn't check the bucket region, but relies on url from the specified region.
So basically in order to this line of code to work:
https://github.com/mlflow/mlflow/blob/c2f76f1aff4f12b80456c36427d8b2f55a8e5d32/mlflow/sagemaker/__init__.py#L599
the client initialisation:
`sess = boto3.Session(region_name=region_name)`
 needs to be in the same region as bucket:
```
s3 = boto3.resource('s3')
bucket_region = s3.get_bucket_location(Bucket=bucket)['LocationConstraint']
if bucket_region:
    sess = boto3.Session(region_name = bucket_region) 
else:
    sess = boto3.Session()
```

**Comment by smurching** [comment]
@Tradunsky thanks for raising this & for the investigation! Is it possible to reproduce this issue using newer versions of MLflow (e.g. MLflow 1.7.0)? The code path you linked hasn't changed, so I'd guess the answer is yes (but if upgrading fixes your issue, that'd also be great to know). Assuming the bug is still present in the most recent release, would you be interested in filing a PR with a fix? Seems the right fix is to ensure Sagemaker fetches the model from the appropriate bucket.

**Comment by Tradunsky** [comment]
Hi @smurching,
Yes, upgrade of mlflow took big effort for a neighbourhood team. Based on my experience with different versions of mlflow, it most likely because some API were deprecated, so we need to migrate all of our services.

Yes, I would be able to create a PR in a week I guess. Just need to unsure I have everything for testing on latest version of the code, since that only available for reproducing on 0.8.1 and 0.9.1.

**Comment by pogil** [comment]
Hi @Tradunsky, have you had an opportunity to follow up on this? We would definitely welcome a contribution here.

**Comment by Tradunsky** [comment]
Hi @pogil, 
Unfortunately, haven't got required environment for reproducing it on latest mlflow version.

I hope I will start working on it in near two weekends.

**Comment by samuelthan** [comment]
We've encounter the same error message ```Failure reason
Failed to download model data for container "container_1" from URL: "s3://xxx-ap-southeast-2-xxx-model/xxxx/1385/model.tar.gz". Please ensure that there is an object located at the URL and that the role passed to CreateModel has permissions to download the object.```.  Found out that if you are encrypting the model using your own custom KMS encryption key, you need to enable kms:decrypt for that specific key in the IAM role used to execute the endpoint creation.  Hope this helps whoever that may still encounter this error

**Comment by dbczumar** [comment]
Thank you for providing a detailed & helpful solution, @samuelthan. I'm going to go ahead and close this issue for now. Feel free to reopen it if the stated problem continues to occur.
